crates/debugging/src/trace/function/trace.rs

```diff
@@ -0,0 +1,131 @@
+use crate::contracts_data_store::ContractsDataStore;
+use crate::trace::function::name::FunctionName;
+use crate::trace::function::node::FunctionNode;
+use crate::trace::function::stack::CallStack;
+use cairo_annotations::map_pcs_to_sierra_statement_ids;
+use cairo_annotations::trace_data::{CasmLevelInfo, TraceEntry};
+use cairo_lang_sierra::extensions::core::{CoreConcreteLibfunc, CoreLibfunc, CoreType};
+use cairo_lang_sierra::program::{GenStatement, StatementIdx};
+use cairo_lang_sierra::program_registry::ProgramRegistry;
+use cheatnet::state::CallTrace;
+use starknet_api::core::ClassHash;
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum FunctionTraceError {
+    #[error("function trace is not supported for forked contracts")]
+    ForkContract,
+}
+/// `FunctionTrace` represents a trace of function calls in a Sierra program.
+/// It captures the structure of function calls and returns, allowing for analysis of the
+/// execution flow within a contract.
+#[derive(Debug, Clone)]
+pub struct FunctionTrace {
+    pub root: FunctionNode,
+}
+
+impl FunctionTrace {
+    /// Creates a new [`FunctionTrace`] from the given [`ClassHash`], [`CallTrace`], and [`ContractsDataStore`].
+    pub fn create(
+        class_hash: ClassHash,
+        call_trace: &CallTrace,
+        contracts_data_store: &ContractsDataStore,
+    ) -> Result<Self, FunctionTraceError> {
+        if contracts_data_store.is_fork(&class_hash) {
+            return Err(FunctionTraceError::ForkContract);
+        }
+        let program_artifact = contracts_data_store
+            .get_program_artifact(&class_hash)
+            .expect("program artifact should be present for local contracts");
+
+        let program = &program_artifact.program;
+
+        let sierra_program_registry = ProgramRegistry::<CoreType, CoreLibfunc>::new(program)
```

Comment:
Some of these objects, like `ProgramRegistry` we already create during forge runtime. I wonder if there would be any significant benefit in trying to reuse them?

Reply:
Same goes for constructing trace